### PR TITLE
Converting the old docstrings to numpydoc style and hiding the license from rendering in every file

### PR DESCRIPTION
**Pull request by George-Elhamy**
## Description
These changes are mainly on converting the already found docstrings in src into numpydoc style which is more readable and better looking when rendered in sphinx and this has been done in two different commits. Along with that, the license have also been hidden from rendering (it still shows in the code in every rst file in the src & it also still shows in the documentation but only in the license section) and this has been done on two commits. There is also two python scripts included in the first commit that claculate the documentation coverage in the whole project so that we can track the coverage as we go on with adding the missing documentation for methods and classes. 

## Types of changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change (Causes existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] I have documented my changes in `/docs/changelog.md`.
- [ ] My change requires additional/updated documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added/updated the config example for any module docstrings as necessary.
- [x] I have checked that the change does not contain obvious errors
(syntax, indentation, mutable default values, etc.).
- [ ] I have tested my changes using 'Load all modules' on the default dummy configuration.
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

### CODE REVIEW

**Comment by timoML** [comment]
To see the built version:
https://qudi-core-testing.readthedocs.io/en/george/index.html

**Review by Neverhorst — CHANGES_REQUESTED** [review]
- In my understanding the leading `..` is not necessary for each line of docstrings and should be removed (except for the very few places where it might be needed).
- Some leftovers from merging conflicts are still in the code
- Indentation is broken for a lot of method docstrings
- Keep single quote string literals to not clutter the commit log and keep it uniform. I stopped marking them at some point because roughly half of the changes in this PR stem from this issue.
- New files need copyright and license disclaimers
- Merge `pyproject.toml` with current `main`. Seems like it is rewriting everything
- Remove `src/qudi/util/uic.py.patch` from this PR
- Remove `src/qudi/util/units.py.patch` from this PR

**Comment by prithviulm** [comment]
Sorry for again making numerous changes and for the force push. This was a mess unfortunately. Hopefully the PR is easier to review now. I've tried to remove all the double quote changes and other formatting changes and other mistakes in the docstrings and licenses etc.

File: src/qudi/core/parentpoller.py [hunk 1]
```diff
@@ -80,14 +81,17 @@ class ParentPollerWindows(Thread):
...
+        quit_function : callable
+            Function to call for shutdown if the parent process is dead.
+        parent_handle : int
+            The program will terminate immediately when this handle is signaled.
         """
         if quit_function is None:
             pass
         elif not callable(quit_function):
-            raise TypeError('argument quit_function must be a callable.')
+            raise TypeError("argument quit_function must be a callable.")
```
Comment (timoML) [inline src/qudi/core/parentpoller.py:94]: unecessary change of string literal.

File: src/qudi/core/statusvariable.py [hunk 2]
```diff
@@ -74,24 +83,38 @@ def copy(self, **kwargs):
...
+        callable
+            The original function so this can be used as a decorator.
         """
         self.representer_function = self._assert_func_signature(func)
         return func
 
     @staticmethod
     def _assert_func_signature(func: Callable) -> Callable:
-        assert callable(func), 'StatusVar constructor/representer must be callable'
+        assert callable(func), "StatusVar constructor/representer must be callable"
```
Comment (timoML) [inline src/qudi/core/statusvariable.py:117]: Unecessary change of string literal.

File: src/qudi/util/datastorage.py [hunk 3]
```diff
@@ -44,20 +44,28 @@
...
-    """
-    PNG = '.png'
-    PDF = '.pdf'
-
+    """Image format to use for saving data thumbnails."""
 
-def get_timestamp_filename(timestamp, nametag=None):
-    """ Returns a qudi standard filename used for saving measurement data to file.
-    Not including any file extension.
+    PNG = ".png"
```
Comment (timoML) [inline src/qudi/util/datastorage.py:49]: Unecessary change of string literal.

File: src/qudi/util/datastorage.py [hunk 4]
```diff
@@ -362,26 +393,52 @@ def save_data(self, data, *, metadata=None, notes=None, nametag=None, timestamp=
...
+            Explicit filename to use for saving the data.
+
+        Returns
+        -------
+        tuple
+            Full file path (str), timestamp used (datetime.datetime), saved data shape (tuple).
         """
-        pass
+
+    pass
```
Comment (timoML) [inline src/qudi/util/datastorage.py:419]: Intendation error will brake function.

File: src/qudi/util/linear_transform.py [hunk 5]
```diff
@@ -263,29 +271,52 @@ def find_changing_axes(points: np.ndarray) -> np.ndarray:
...
 
-    :param v0: Support vector 0
-    :param v1: Support vector 1
-    :param v2: Support vector 2
-    :param ez: Normal vector defining the target plane. By default this is the z plane.
-    :return: 3x3 rotation matrix
     """
     if len(v0) != 3 or len(v1) != 3 or len(v2) != 3:
-        raise ValueError('The support vectors should have a length of 3.')
+        raise ValueError("The support vectors should have a length of 3.")
```
Comment (timoML) [inline src/qudi/util/linear_transform.py:319]: Unecessary change of string literal.

File: src/qudi/util/mutex.py [hunk 6]
```diff
@@ -40,15 +39,22 @@ class Mutex(_QMutex):
     * Context management (enter/exit)
     """
 
-    def acquire(self, blocking: Optional[bool] = True, timeout: Optional[_RealNumber] = -1) -> bool:
-        """ Mimics threading.Lock.acquire() to allow this class as a drop-in replacement.
+    def acquire(
```
Comment (timoML) [inline src/qudi/util/mutex.py:42]: function signature does not follow project standards

File: src/qudi/util/mutex.py [hunk 7]
```diff
@@ -91,16 +103,27 @@ class RecursiveMutex(_QRecursiveMutex):
         refactoring your code to use a simple mutex before using this object.
         """
 
-        def acquire(self, blocking: Optional[bool] = True,
-                    timeout: Optional[_RealNumber] = -1) -> bool:
-            """ Mimics threading.Lock.acquire() to allow this class as a drop-in replacement.
+        def acquire(
```
Comment (timoML) [inline src/qudi/util/mutex.py:106]: function signature does not follow project standards

File: src/qudi/util/uic.py [hunk 8]
```diff
@@ -49,8 +49,20 @@ def loadUi(file_path, base_widget):
...
 
-    @param str file_path: The full path to the .ui-file to load
-    @param object base_widget: Instance of the base widget represented by the .ui-file
+    Parameters
+    ----------
+    str :
+        file_path: The full path to the .ui-file to load
+    object :
+        base_widget: Instance of the base widget represented by the .ui-file
+    file_path :
```
Comment (timoML) [inline src/qudi/util/uic.py:58]: docstring seems broken

File: src/qudi/util/uic.py [hunk 9]
```diff
@@ -107,8 +119,18 @@ def _convert_ui_to_absolute_paths(file_path):
     """ Converts the .ui file in order to change all relative path declarations containing the
     keyword "/artwork/" into absolute paths pointing to the qudi artwork data directory.
 
-    @param str file_path: The path to the .ui file to convert
-    @return str|NoneType: Converted file content of the .ui file, None if conversion is not needed
+    Parameters
+    ----------
+    str :
+        file_path: The path to the .ui file to convert
+    file_path :
```
Comment (timoML) [inline src/qudi/util/uic.py:126]: docstring seems broken

File: src/qudi/util/uic.py [hunk 10]
```diff
@@ -107,8 +119,18 @@ def _convert_ui_to_absolute_paths(file_path):
...
+        
+
+    Returns
+    -------
+    type
+        str|NoneType: Converted file content of the .ui file, None if conversion is not needed
```
Comment (timoML) [inline src/qudi/util/uic.py:132]: type definition seems broken

**Comment by timoML** [comment]
- I think Sphinx files (Makefile, conf.py) should go into a seperate folder docs/sphinx
- Did we want to keep string literals as they were or change to consistent use?

**Comment by timoML** [comment]
Runs on dummy with "load all modules". So indentation in (non-docstring) code seems fine.

**Comment by Neverhorst** [comment]
Is the double-dot `..` in the license/copyright disclaimer needed? Seems weird to introduce this in all files since it is not required for a simple text body as far as I understand.

**Comment by timoML** [comment]
It's indicating to sphinx to not render the license for the documentation of every class.
Is there another way for achieving the same?

**Comment by Neverhorst** [comment]
> It's indicating to sphinx to not render the license for the documentation of every class. Is there another way for achieving the same?

Not from the top of my head but this documentation on [docstring preprocessing](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#docstring-preprocessing) seems promising.

**Comment by prithviulm** [comment]
It assumes that text at the top of the file is the module docstring and so the license also gets rendered along with the module docstring (if any). There might be some kind of workaround using regex to intercept the text and remove the license after Sphinx reads the text but before it starts parsing.

**Comment by prithviulm** [comment]
Ok, I think this is ready to merge now. Fixed a few more docstrings that were not converted. Reverted the license back to remove the leading `..`. As discussed, a style guide is necessary for things like when and where to use newlines in docstrings as Ruff will not handle these. I will add a brief description of it to the Ruff config.

**Review by timoML — APPROVED** [review]
Looks good now, thanks for the effort.

**Comment by prithviulm** [comment]
Check that it still runs on dummy with "load all modules".

**Comment by prithviulm** [comment]
@Neverhorst We've addressed the remaining changes but can't merge till you approve.

**Review by Neverhorst — APPROVED** [review]
Thank you for your effort @prithviulm . Looks very good now.
Merging the testing and documentation overhaul "one-thing-at-a-time" makes this way more manageable.

**Pull request merged**